File: python_backend/utils/logger.py
import logging
import sys
from datetime import datetime
from pathlib import Path

# Ensure logs directory exists
log_dir = Path(__file__).parent.parent / 'logs'
log_dir.mkdir(exist_ok=True)

# Custom formatter with JSON-like structure
class StructuredFormatter(logging.Formatter):
    def format(self, record):
        log_data = {
            'timestamp': datetime.utcnow().isoformat(),
            'level': record.levelname,
            'logger': record.name,
            'message': record.getMessage(),
            'module': record.module,
            'function': record.funcName,
            'line': record.lineno,
        }
        
        # Add exception info if present
        if record.exc_info:
            log_data['exception'] = self.formatException(record.exc_info)
        
        # Add extra fields if present
        if hasattr(record, 'user_id'):
            log_data['user_id'] = record.user_id
        if hasattr(record, 'tenant_id'):
            log_data['tenant_id'] = record.tenant_id
        
        return str(log_data)

# Configure root logger
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s',
    handlers=[
        logging.FileHandler(log_dir / 'app.log'),
        logging.StreamHandler(sys.stdout)
    ]
)

# Get logger instance
logger = logging.getLogger('storeai_python')

# Add structured file handler
structured_handler = logging.FileHandler(log_dir / 'structured.log')
structured_handler.setFormatter(StructuredFormatter())
logger.addHandler(structured_handler)

# Helper functions
def log_error(error: Exception, context: dict = None):
    """Log an error with context"""
    extra = context or {}
    logger.error(f"Error: {str(error)}", exc_info=True, extra=extra)

def log_query(query_type: str, query: str, duration: float, error: Exception = None):
    """Log a database query"""
    if error:
        logger.error(
            f"Query failed: {query_type}",
            extra={'query': query[:200], 'duration': duration, 'error': str(error)}
        )
    else:
        logger.debug(
            f"Query executed: {query_type}",
            extra={'query': query[:200], 'duration': duration}
        )

def log_api_call(method: str, path: str, user_id: str = None, tenant_id: str = None):
    """Log an API call"""
    logger.info(
        f"API {method} {path}",
        extra={'user_id': user_id, 'tenant_id': tenant_id}
    )

__all__ = ['logger', 'log_error', 'log_query', 'log_api_call']

File: main/server/src/utils/logger.ts
import winston from 'winston';
import path from 'path';

// Ensure logs directory exists
const logsDir = path.join(__dirname, '../../logs');

// Custom format for console output (human-readable)
const consoleFormat = winston.format.combine(
    winston.format.colorize(),
    winston.format.timestamp({ format: 'YYYY-MM-DD HH:mm:ss' }),
    winston.format.printf(({ timestamp, level, message, ...meta }) => {
        let msg = `${timestamp} [${level}]: ${message}`;
        if (Object.keys(meta).length > 0) {
            msg += ` ${JSON.stringify(meta)}`;
        }
        return msg;
    })
);

// Structured format for file output (JSON for parsing)
const fileFormat = winston.format.combine(
    winston.format.timestamp(),
    winston.format.errors({ stack: true }),
    winston.format.json()
);

const isServerless = process.env.VERCEL === '1';
const transports: winston.transport[] = [
    new winston.transports.Console({
        format: consoleFormat,
        level: process.env.NODE_ENV === 'production' ? 'warn' : 'debug',
    }),
];

// Vercel function filesystems are read-only. Persist logs only on long-lived
// server processes and send serverless logs to Vercel's console collector.
if (!isServerless) {
    transports.unshift(
        new winston.transports.File({
            filename: path.join(logsDir, 'error.log'),
            level: 'error',
            format: fileFormat,
            maxsize: 5242880,
            maxFiles: 5,
        }),
        new winston.transports.File({
            filename: path.join(logsDir, 'combined.log'),
            format: fileFormat,
            maxsize: 5242880,
            maxFiles: 10,
        })
    );
}

// Create logger instance
export const logger = winston.createLogger({
    level: process.env.LOG_LEVEL || 'info',
    defaultMeta: { service: 'storeai-backend' },
    transports,
});

// Stream for Morgan HTTP logging
export const httpLogStream = {
    write: (message: string) => {
        logger.http(message.trim());
    },
};

// Helper functions for common logging patterns
export const logError = (error: Error, context?: Record<string, any>) => {
    logger.error('Error occurred', {
        message: error.message,
        stack: error.stack,
        ...context,
    });
};

export const logAPICall = (method: string, path: string, userId?: string, tenantId?: string) => {
    logger.info('API Call', {
        method,
        path,
        userId,
        tenantId,
        timestamp: new Date().toISOString(),
    });
};

export const logDatabaseQuery = (query: string, duration: number, error?: Error) => {
    if (error) {
        logger.error('Database query failed', {
            query: query.substring(0, 200), // Limit query length
            duration,
            error: error.message,
        });
    } else {
        logger.debug('Database query executed', {
            query: query.substring(0, 200),
            duration,
        });
    }
};

export default logger;
